text/0048-session-keys-runtime-api.md

```diff
@@ -0,0 +1,115 @@
+key signing the `account_id`. The actual type of each signature depends on the
+corresponding session key cryptographic algorithm. The order of the signatures in 
+the `proof` is the same as the order of the session keys in the `SessionKeys` type 
+declared in the runtime.
+
+The version of the `SessionKeys` needs to be bumped to `1` to reflect the changes to the 
+signature of `SessionKeys_generate_session_keys`:
+```rust
+pub struct OpaqueGeneratedSessionKeys {
+	pub keys: Vec<u8>,
+	pub proof: Vec<u8>,
+}
+
+fn SessionKeys_generate_session_keys(account_id: Vec<u8>, seed: Option<Vec<u8>>) -> OpaqueGeneratedSessionKeys;
+```
+
+The default calling convention for runtime apis is applied, meaning the parameters 
+passed as SCALE encoded array and the length of the encoded array. The return value 
+being the SCALE encoded return value as `u64` (`array_ptr | length << 32`). So, the 
+actual exported function signature looks like:
+```rust
+fn SessionKeys_generate_session_keys(array: *const u8, len: usize) -> u64;
+```
+
+The on chain logic for setting the `SessionKeys` needs to be changed as well. It
+already gets the `proof` passed as `Vec<u8>`. This `proof` needs to be decoded to
+the actual `Proof` type as explained above. The `proof` and the SCALE encoded
+`account_id` of the sender are used to verify the ownership of the `SessionKeys`.
+
+## Drawbacks
+
+Validator operators need to pass the their account id when rotating their session keys in a node. 
+This will require updating some high level docs and making users familiar with the slightly changed ergonomics.
+
+## Testing, Security, and Privacy
+
+Testing of the new changes only requires passing an appropriate `owner` for the current testing context. 
+The changes to the proof generation and verification got audited to ensure they are correct.
+
+## Performance, Ergonomics, and Compatibility
+
+### Performance
+
+The session key generation is an offchain process and thus, doesn't influence the performance of the 
+chain. Verifying the proof is done on chain as part of the transaction logic for setting the session keys. 
+The verification of the proof is a signature verification number of individual session keys times. As setting
+the session keys is happening quite rarely, it should not influence the overall system performance.
+
+### Ergonomics
+
+The interfaces have been optimized to make it as easy as possible to generate the ownership proof.
+
+### Compatibility
+
+Introduces a new version of the `SessionKeys` runtime api. Thus, nodes should be updated before 
+a runtime is enacted that contains these changes otherwise they will fail to generate session keys. 
+The RPC that exists around this runtime api needs to be updated to support passing the account id 
+and for returning the ownership proof alongside the public session keys.
+
+UIs would need to be updated to support the new RPC and the changed on chain logic.
```

Comment:
There is a new RPC for this. So, they will not break. However when the runtime is upgraded it will require the proofs which only the new UI generates. There is also only polkadot-js which supports this AFAIK.